README.md
Carver Bartz  cjbartz@iastate.edu
Dylan Hampton dhampton@iastate.edu
Ben Nguyen    btnguyen@iastate.edu

This program generates a basic dungeon model according to the assignment 1.09 specifications. We created methods in pc.cpp, to remove, expunge, grab, wear, drop items. We also limited the items in a player's inventory to be only one of each type, aside from rings where you can wear 2 of them. Updated to combat to make the only win condition to kill a monster with the boss tag, and combat is now damage based. every attack is not an instant kill anymore, instead it rolls the damage dice, and subtracts that from the defender's health. We made it so that you can look at item and monster descriptions. also you can list all of the items and equipment on the player.


CHANGELOG
4/15/21 BN/DH: fixed buggy equipment limiting code, updated player stats when equipping items/removing items
4/14/21 CB: Limited equipment based off type
4/14/21 BN/DH: made inspect monster method, implemented damage and hp based combat, made so monsters can't kill eachother
4/13/21 BN/DH: fixed bug where we couldn't see items after we placed them somewhere else, also made expunge, remove and wear item methods, made pc item methods that needed slots, amde those work, added colors to the menus, added killing spongebob as the win condition
4/13/21 BN/DH: made displays for player inventory and gear, fixed duplication bug, but dropping items now sometimes doesn't display
4/12/21 BN/DH/CB: in pc.cpp we made pc_grab_item that picks up the item and puts it into inv, testing phase for all item management methods, need to fix duping bug with dropping items
4/8/21 BN/DH: Made the no fog show all monsters and items in color
4/7/21 BN/DH: Finished object generation and deallocated the objects after the dungeon is deleted
4/6/21 BN/DH: Made monster generation from vectors and started object generation, added color printing
4/3/21 BN/DH: changed monster generation to be from desc, added monster factory
3/30/21 BN/DH: Finished abilities, desc, added dice and description classes, added print function for parser //1.07 assignment done 
3/29/21 BN/DH/CB: Finished color, damage, hp, and speed. Still need to do description
3/24/21 BN: fixed the bug where you could teleport onto yourself
3/23/21 BN/DH/CB: finished everything for 1.06 (fixed fog of war, fixed teleporting onto a monster, implemented random teleportation) //1.06 assignment done
3/22/21 BN/DH/CB: Worked on telport mechanic, finished targeting mode, sometimes Z's show up in the fog of war mode and we don't know why, also need to fix teleporting onto a monster
3/21/21 CB: Started the teleportation mechanic, just the random.
3/20/21 BN/DH: Changed all .c files (except for heap.c) to .cpp, fixed any errors, added fog of war, added f key, which shows the entire screen
3/11/21 BN: Deleted save and load functions, and added comments to the main game loop
3/09/21 BN/DH: Updated scroll for monster list and added prefixes to the monster names, fixed bug where number of monsters in monster list didn't match alive monsters. //1.05 assignment done
3/09/21 BN: Added so that if there is an invalid input, monsters will not move, made it so it updates the amount of data allocated for monster_list in update_monster_list
3/09/21 DH: Added scroll functionality to monster list
3/08/21 DH/BN/CB: Added reserve for line 0 for text messages, added monster display
3/06/21 DH/BN: Added EPIC player win screen, added player_movement.c which contains both methods to move player, stopped monsters from moving onto stairs, added update_monster_list
3/05/2021 BN: Added spawn_new_dungeon method to make a new dungeon without having to call a million methods. also implemented ncurses version print_dungeon, made the player movement fully functional, added stairs functionality (does not remember previous dungeons), added quit button
3/03/2021 DH/BN: Added corridor creation when monster moves through rock, added monster deaths, added player lose case, refactored variable names, bug fixed, and added EPIC ending screen, added erratic and dumb characterisitic //1.04 assignment done
3/02/2021 DH/BN: modified entities to be based on 2d array, modified print statement, fixed heap initialization and next_turn, got telepathic monsters to move the way they are supposed to
3/01/2021 DH/BN/CB: Added randomness to monster generation and added generate_entities_heap, started on next_turn
2/28/2021 DH/BN: Added create_entities method and added monster generation
2/22/2021 DH/BN/CB: Added print distance maps function in dungeon.c, and bugtested //1.03 assignment done
2/21/2021 DH/BN: Implemented distance map in path_finding.c for both tunneling and non-tunneling monsters
2/21/2021 DH/BN: Added dungeon.h, added path_finding.c, heap.c, heap.h
2/17/2021 DH/BN/CB: Added endianess //1.02 assignment done
2/17/2021 BN: fixed saving and loading the arrays, and fixed loading rooms from a file
2/17/2021 DH/BN/CB: Worked on correcting load_dungeon
2/16/2021 DH: Completed save dungeon, working on tracing segfaults in read dungeon
2/15/2021 DH/BN/CB: Worked on getting save dungeon complete, still need to trace some seg faults
2/15/2021 DH: Added PC struct and fixed some bugs with save_dungeon, added up to num_rooms in load_dungeon
2/14/2021 DH: Restructured room code in favor of using a struct and added dungeon_hardness and num_rooms writes in save_dungeon
2/13/2021 DH: Added parallel 2d array for storing hardness factor
2/12/2021 BN: Added create_player and cleaned up save and load tag setting
2/11/2021 DH: Added save / load tag checks and set up save_dungeon and load_dungeon skeletons
2/8/2021 BN/DH: Added simple path generation // 1.01 assignment done
2/5/2021 BN: Added stair generation and fixed some bugs
2/5/2021 DH: Added a check to avoid two rooms touching
2/4/2021 DH: Created Makefile and changed formatting.
2/4/2021 CB: Set up code, made create_rooms method, set_dungeon method, and
print_dungeon method.



move.cpp
#include "move.h"

#include <unistd.h>
#include <stdlib.h>
#include <assert.h>

#include "dungeon.h"
#include "heap.h"
#include "move.h"
#include "npc.h"
#include "pc.h"
#include "character.h"
#include "utils.h"
#include "path.h"
#include "event.h"
#include "io.h"

void do_combat(dungeon *d, character *atk, character *def)
{
  int can_see_atk, can_see_def;
  const char *organs[] = {
    "liver",                   /*  0 */
    "pancreas",                /*  1 */
    "heart",                   /*  2 */
    "eye",                     /*  3 */
    "arm",                     /*  4 */
    "leg",                     /*  5 */
    "intestines",              /*  6 */
    "gall bladder",            /*  7 */
    "lungs",                   /*  8 */
    "hand",                    /*  9 */
    "foot",                    /* 10 */
    "spinal cord",             /* 11 */
    "pituitary gland",         /* 12 */
    "thyroid",                 /* 13 */
    "tongue",                  /* 14 */
    "bladder",                 /* 15 */
    "diaphram",                /* 16 */
    "stomach",                 /* 17 */
    "pharynx",                 /* 18 */
    "esophagus",               /* 19 */
    "trachea",                 /* 20 */
    "urethra",                 /* 21 */
    "spleen",                  /* 22 */
    "ganglia",                 /* 23 */
    "ear",                     /* 24 */
    "subcutaneous tissue",      /* 25 */
    "cerebellum",              /* 26 */ /* Brain parts begin here */
    "hippocampus",             /* 27 */
    "frontal lobe",            /* 28 */
    "brain",                   /* 29 */
  };
  int part;
  int attack_roll = atk->damage.roll();

  if (def->alive && (atk == d->PC || def == d->PC)) {
    if (attack_roll >= def->hitpoints) {
      def->alive = 0;
      charpair(def->position) = NULL;

      if (def != d->PC) {
        d->num_monsters--;
      } else {
        if ((part = rand() % (sizeof (organs) / sizeof (organs[0]))) < 26) {
          io_queue_message("As the %c eats your %s, "
              "you wonder if there is an afterlife.",
              atk->symbol, organs[part]);
        } else {
          io_queue_message("Your last thoughts fade away as "
              "the %c eats your %s...",
              atk->symbol, organs[part]);
        }
        /* Queue an empty message, otherwise the game will not pause for *
         * player to see above.                                          */
        io_queue_message("");
      }
      atk->kills[kill_direct]++;
      atk->kills[kill_avenged] += (def->kills[kill_direct] +
          def->kills[kill_avenged]);
      if (((npc *) def)->characteristics & NPC_BOSS && atk == d->PC) {
        d->PC->boss_kills++;
      }

      can_see_atk = can_see(d, character_get_pos(d->PC),
          character_get_pos(atk), 1, 0);
      can_see_def = can_see(d, character_get_pos(d->PC),
          character_get_pos(def), 1, 0);

      if (atk != d->PC && def != d->PC) {
        if (can_see_atk && !can_see_def) {
          io_queue_message("The %c callously murders some poor, "
              "defenseless creature.", atk->symbol);
        }
        if (can_see_def && !can_see_atk) {
          io_queue_message("Something kills the helpless %c.", def->symbol);
        }
        if (can_see_atk && can_see_def) {
          io_queue_message("You watch in abject horror as the %c "
              "gruesomely murders the %c!", atk->symbol, def->symbol);
        }
      }
    }

    if (atk == d->PC) {
      io_queue_message("You smite the %c for %d damage!", def->symbol, attack_roll);
      def->hitpoints -= attack_roll;
    }
  }
}

void move_character(dungeon *d, character *c, pair_t next)
{
  if (charpair(next) &&
      ((next[dim_y] != c->position[dim_y]) ||
       (next[dim_x] != c->position[dim_x]))) {
    do_combat(d, c, charpair(next));
  } else {
    /* No character in new position. */

    d->character_map[c->position[dim_y]][c->position[dim_x]] = NULL;
    c->position[dim_y] = next[dim_y];
    c->position[dim_x] = next[dim_x];
    d->character_map[c->position[dim_y]][c->position[dim_x]] = c;
  }

  if (c == d->PC) {
    pc_reset_visibility(d->PC);
    pc_observe_terrain(d->PC, d);
  }
}

void do_moves(dungeon *d)
{
  pair_t next;
  character *c;
  event *e;

  /* Remove the PC when it is PC turn.  Replace on next call.  This allows *
   * use to completely uninit the heap when generating a new level without *
   * worrying about deleting the PC.                                       */

  if (pc_is_alive(d)) {
    /* The PC always goes first one a tie, so we don't use new_event().  *
     * We generate one manually so that we can set the PC sequence       *
     * number to zero.                                                   */
    e = (event *) malloc(sizeof (*e));
    e->type = event_character_turn;
    /* Hack: New dungeons are marked.  Unmark and ensure PC goes at d->time, *
     * otherwise, monsters get a turn before the PC.                         */
    if (d->is_new) {
      d->is_new = 0;
      e->time = d->time;
    } else {
      e->time = d->time + (1000 / d->PC->speed);
    }
    e->sequence = 0;
    e->c = d->PC;
    heap_insert(&d->events, e);
    pc_grab_item(d->PC, d);
  }

  while (pc_is_alive(d) &&
      (e = (event *) heap_remove_min(&d->events)) &&
      ((e->type != event_character_turn) || (e->c != d->PC))) {
    d->time = e->time;
    if (e->type == event_character_turn) {
      c = e->c;
    }
    if (!c->alive) {
      if (d->character_map[c->position[dim_y]][c->position[dim_x]] == c) {
        d->character_map[c->position[dim_y]][c->position[dim_x]] = NULL;
      }
      if (c != d->PC) {
        event_delete(e);
      }
      continue;
    }

    npc_next_pos(d, (npc *) c, next);
    move_character(d, c, next);

    heap_insert(&d->events, update_event(d, e, 1000 / c->speed));
  }

  io_display(d);
  if (pc_is_alive(d) && e->c == d->PC) {
    c = e->c;
    d->time = e->time;
    /* Kind of kludgey, but because the PC is never in the queue when   *
     * we are outside of this function, the PC event has to get deleted *
     * and recreated every time we leave and re-enter this function.    */
    e->c = NULL;
    event_delete(e);
    io_handle_input(d);
  }
}

void dir_nearest_wall(dungeon *d, character *c, pair_t dir)
{
  dir[dim_x] = dir[dim_y] = 0;

  if (c->position[dim_x] != 1 && c->position[dim_x] != DUNGEON_X - 2) {
    dir[dim_x] = (c->position[dim_x] > DUNGEON_X - c->position[dim_x] ? 1 : -1);
  }
  if (c->position[dim_y] != 1 && c->position[dim_y] != DUNGEON_Y - 2) {
    dir[dim_y] = (c->position[dim_y] > DUNGEON_Y - c->position[dim_y] ? 1 : -1);
  }
}

uint32_t against_wall(dungeon *d, character *c)
{
  return ((mapxy(c->position[dim_x] - 1,
          c->position[dim_y]    ) == ter_wall_immutable) ||
      (mapxy(c->position[dim_x] + 1,
             c->position[dim_y]    ) == ter_wall_immutable) ||
      (mapxy(c->position[dim_x]    ,
             c->position[dim_y] - 1) == ter_wall_immutable) ||
      (mapxy(c->position[dim_x]    ,
             c->position[dim_y] + 1) == ter_wall_immutable));
}

uint32_t in_corner(dungeon *d, character *c)
{
  uint32_t num_immutable;

  num_immutable = 0;

  num_immutable += (mapxy(c->position[dim_x] - 1,
        c->position[dim_y]    ) == ter_wall_immutable);
  num_immutable += (mapxy(c->position[dim_x] + 1,
        c->position[dim_y]    ) == ter_wall_immutable);
  num_immutable += (mapxy(c->position[dim_x]    ,
        c->position[dim_y] - 1) == ter_wall_immutable);
  num_immutable += (mapxy(c->position[dim_x]    ,
        c->position[dim_y] + 1) == ter_wall_immutable);

  return num_immutable > 1;
}

static void new_dungeon_level(dungeon *d, uint32_t dir)
{
  /* Eventually up and down will be independantly meaningful. *
   * For now, simply generate a new dungeon.                  */

  switch (dir) {
    case '<':
    case '>':
      new_dungeon(d);
      break;
    default:
      break;
  }
}


uint32_t move_pc(dungeon *d, uint32_t dir)
{
  pair_t next;
  uint32_t was_stairs = 0;
  const char *wallmsg[] = {
    "There's a wall in the way.",
    "BUMP!",
    "Ouch!",
    "You stub your toe.",
    "You can't go that way.",
    "You admire the engravings.",
    "Are you drunk?"
  };

  next[dim_y] = d->PC->position[dim_y];
  next[dim_x] = d->PC->position[dim_x];


  switch (dir) {
    case 1:
    case 2:
    case 3:
      next[dim_y]++;
      break;
    case 4:
    case 5:
    case 6:
      break;
    case 7:
    case 8:
    case 9:
      next[dim_y]--;
      break;
  }
  switch (dir) {
    case 1:
    case 4:
    case 7:
      next[dim_x]--;
      break;
    case 2:
    case 5:
    case 8:
      break;
    case 3:
    case 6:
    case 9:
      next[dim_x]++;
      break;
    case '<':
      if (mappair(d->PC->position) == ter_stairs_up) {
        was_stairs = 1;
        new_dungeon_level(d, '<');
      }
      break;
    case '>':
      if (mappair(d->PC->position) == ter_stairs_down) {
        was_stairs = 1;
        new_dungeon_level(d, '>');
      }
      break;
  }

  if (was_stairs) {
    return 0;
  }

  if ((dir != '>') && (dir != '<') && (mappair(next) >= ter_floor)) {
    move_character(d, d->PC, next);
    dijkstra(d);
    dijkstra_tunnel(d);

    return 0;
  } else if (mappair(next) < ter_floor) {
    io_queue_message(wallmsg[rand() % (sizeof (wallmsg) /
          sizeof (wallmsg[0]))]);
    io_display(d);
  }

  return 1;
}
